### PR TITLE
C++17: std::char_traits<>::{compare,length} is constexpr.

### PR DESCRIPTION
**Pull request by phprus**
Discussion: https://github.com/fmtlib/fmt/pull/2243#discussion_r616803322

### CODE REVIEW

File: include/fmt/format.h [hunk 1]
```diff
@@ -2149,7 +2149,10 @@ FMT_CONSTEXPR OutputIt write(OutputIt out, Char value) {
 }
 
 template <typename Char, typename OutputIt>
-FMT_CONSTEXPR OutputIt write(OutputIt out, const Char* value) {
+#if FMT_HAS_CONSTEXPR_CHAR_TRAITS  // C++17's char_traits::length() is constexpr.
+FMT_CONSTEXPR
```
Comment (alexezeder) [inline include/fmt/format.h:2153]: No need to use `FMT_CONSTEXPR` macro here, just `constexpr` would be enough.

File: include/fmt/core.h [hunk 2]
```diff
@@ -457,6 +468,9 @@ template <typename Char> class basic_string_view {
   }
 
   // Lexicographically compare this string reference to other.
+#if FMT_HAS_CONSTEXPR_CHAR_TRAITS  // C++17's char_traits::compare() is constexpr.
+  constexpr
+#endif
```
Comment (alexezeder) [inline include/fmt/core.h:473]: No need to make this function `constexpr` in this PR, I think.
This can be done as needed in the following PRs.

File: include/fmt/color.h [hunk 3]
```diff
@@ -432,7 +432,10 @@ template <typename Char> struct ansi_color_escape {
   FMT_CONSTEXPR operator const Char*() const FMT_NOEXCEPT { return buffer; }
 
   FMT_CONSTEXPR const Char* begin() const FMT_NOEXCEPT { return buffer; }
-  FMT_CONSTEXPR const Char* end() const FMT_NOEXCEPT {
+#if FMT_HAS_CONSTEXPR_CHAR_TRAITS  // C++17's char_traits::length() is constexpr.
+  FMT_CONSTEXPR
```
Comment (alexezeder) [inline include/fmt/color.h:436]: No need to use `FMT_CONSTEXPR` macro here, just `constexpr` would be enough.

File: include/fmt/core.h [hunk 4]
```diff
@@ -43,6 +43,17 @@
...
+#if __cplusplus >= 201703L
+#  ifndef __GLIBCXX__
+#    define FMT_HAS_CONSTEXPR_CHAR_TRAITS 1
+#  elif defined(_GLIBCXX_RELEASE) && _GLIBCXX_RELEASE >= 7
+#    define FMT_HAS_CONSTEXPR_CHAR_TRAITS 1
+#  endif
+#endif
+#ifndef FMT_HAS_CONSTEXPR_CHAR_TRAITS
+#  define FMT_HAS_CONSTEXPR_CHAR_TRAITS 0
+#endif
```
Comment (alexezeder) [inline include/fmt/core.h:55]: I suggest you comment on this check because it's not a really straightforward one 🙂

**Review by alexezeder — CHANGES_REQUESTED** [review]
Just a few suggestions 😉

File: include/fmt/core.h (hunk 2 above)
Comment (phprus) [reply include/fmt/core.h:473]: This change is for consistency across all uses of ``constexpr std::char_traits``.

Comment (alexezeder) [reply include/fmt/core.h:473]: All I'm just saying is that this function is non-`constexpr` currently, unlike `basic_string_view` constructor or `write(const Char* value)`, so there is no need to make it `constexpr` in this PR.

**Review by vitaut — CHANGES_REQUESTED** [review]
Thanks for the PR. In general looks good but let's replace `FMT_HAS_CONSTEXPR_CHAR_TRAITS` with `FMT_CONSTEXPR_CHAR_TRAITS` that expands to `constexpr` or nothing depending on availability of constexpr char traits. This will make usage cleaner which is important because we plan to use this macro in multiple places.

**Comment by vitaut** [comment]
Actually looking at the latest diff, you already made the suggested change =). GitHub confusingly shows the old code in Conversation.

**Review by vitaut — APPROVED** [review]

**Comment by alexezeder** [comment]
It looks like there is a problem here that no one can ever expect: https://godbolt.org/z/Ps7q8dGME
TLDR: `__cplusplus` is `030c1fH = 199711` on MSVC even with `/std:c++17` option, because [it's MSVC](https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/) I guess...
🤦 🤦 🤦

**Comment by phprus** [comment]
Oh!!!
My big mistake...

I can make pull request tonight.

Fix:
```
#if __cplusplus >= 201703L
#  ifndef __GLIBCXX__
#    define FMT_CONSTEXPR_CHAR_TRAITS constexpr
#  elif defined(_GLIBCXX_RELEASE) && _GLIBCXX_RELEASE >= 7
#    define FMT_CONSTEXPR_CHAR_TRAITS constexpr
#  endif
#endif
```
to:
```
#if defined(_MSC_VER)
// VS 2017 15.7+ (https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=msvc-150)
#  if _MSVC_LANG >= 201703L && _MSC_VER >= 1914
#    define FMT_CONSTEXPR_CHAR_TRAITS constexpr
#  endif
#elif __cplusplus >= 201703L
#  if defined(_LIBCPP_VERSION) && _LIBCPP_VERSION >= 4000 
// libc++4.0+ (https://libcxx.llvm.org/docs/Cxx1zStatus.html)
#    define FMT_CONSTEXPR_CHAR_TRAITS constexpr
#  elif defined(__GLIBCXX__) && defined(_GLIBCXX_RELEASE) && _GLIBCXX_RELEASE >= 7
#    define FMT_CONSTEXPR_CHAR_TRAITS constexpr
#  endif
#endif
```

New comment:
```
// Check if constexpr std::char_traits<>::compare,length is supported.
// libstdc++: GCC 7 and newer and __cplusplus >= 201703L
// MSVC: VS 2017 15.7 and newer and /std:c++17
// libc++: 4.0 and newer and if __cplusplus >= 201703L
// NOTE: FMT_GCC_VERSION  - is not libstdc++ version.
//       _GLIBCXX_RELEASE - is present in GCC 7 libstdc++ and newer.
```

**Comment by phprus** [comment]
@alexezeder, Can you make a pull request?

**Comment by alexezeder** [comment]
Actually I already somehow fixed this with https://github.com/fmtlib/fmt/pull/2247/commits/8a03ec0a17aa746b06a2e13f563be05b3443dcea:
```diff
- #if __cplusplus >= 201703L
+ #if __cplusplus >= 201703L || (defined(_MSVC_LANG) && _MSVC_LANG >= 201703L)
```
Looking at your proposal, probably it's not enough since you have an additional `_MSC_VER >= 1914` check there.

> Can you make a pull request?

Of course I can, but it's not needed right now because, as I said, I make it work for my PR. Feel free to make PR at your convenience.

**Comment by phprus** [comment]
```_MSVC_LANG >= 201703L``` - VS2017 version 15.3+
```constexpr char_traits``` - VS 2017 version 15.7+
